lib/fizzy/instantiate.hpp

```diff
@@ -117,11 +117,47 @@ struct ImportedFunction
     std::function<ExecutionResult(Instance&, const Value*, int depth)> function;
 };
 
-// Create vector of ExternalFunctions ready to be passed to instantiate.
-// imported_functions may be in any order,
-// but must contain functions for all of the imported function names defined in the module.
-std::vector<ExternalFunction> resolve_imported_functions(
-    const Module& module, std::vector<ImportedFunction> imported_functions);
+struct ImportedTable
+{
+    std::string module;
+    std::string name;
+    table_elements* table = nullptr;
+    uint32_t min = 0;
+    std::optional<uint32_t> max;
+};
+
+struct ImportedMemory
+{
+    std::string module;
+    std::string name;
+    bytes* data = nullptr;
+    uint32_t min = 0;
+    std::optional<uint32_t> max;
+};
+
+struct ImportedGlobal
+{
+    std::string module;
+    std::string name;
+    Value* value = nullptr;
+    ValType type = ValType::i32;
+    bool is_mutable = false;
+};
+
+struct Externals
+{
+    std::vector<ExternalFunction> functions;
+    std::vector<ExternalTable> tables;
+    std::vector<ExternalMemory> memories;
+    std::vector<ExternalGlobal> globals;
+};
+
+// Create vectors of External* ready to be passed to instantiate.
+// imported_* may be in any order,
+// but must contain imports for all of the import names defined in the module.
+Externals resolve_imports(const Module& module, std::vector<ImportedFunction> imported_functions,
```

Comment:
I think it would be nice to structure this differently, from a consumer side:
```
Externals externals(module);
externals.resolve_global(ExternalGlobal);
externals.resolve_globals(...); // with a span
```

Reply:
At what point would it check that all required imports are provided? In a separate `bool is_complete()` method?

Reply:
I thinks I'm still more in favor of simpler functional API like `vector<External*> resolve_imported_*(Module, span<Imported*>)` because the rest of the API is mostly functional.

(Otherwise it seems better to make this a full-fledged builder object with `instantiate` method in the end, but then there'll probably be some complications around `module` and externals being consumed by `instantiate`...)

I will split `resolve_imports` into 4 functions for now and will try to use `span`.

Reply:
I'm not convinced this is good design, but I don't think this is important right now. Good API is hard to do on paper, so I will accept whatever forks for Rust currently.

Better is to observe how this works in some real-life examples. We have two of these right now: `fizzy-wasi` and `FizzyEngine`.

Reply:
In general I'm not content either with any of the options for resolve proposed so far, and would just pick something good enough. 

(`fizzy-wasi` and `FizzyEngine` are quite simplistic use cases, as they import only functions. I think any of the options would look fine in them)

Reply:
My point is we should start with something "good enough" and the tune it based on example integrations, not on intuition nor unit tests.